From 1d88b0b8d271bde76aba392b276c1d0918665143 Mon Sep 17 00:00:00 2001
From: Hongwei Wang <hwwang@google.com>
Date: Thu, 28 Aug 2025 11:19:07 -0700
Subject: [PATCH 09/55] GRANT TRUSTED_OVERLAY after PiP is entered

- Make sure the TRUSTED_OVERLAY is only granted after WMShell
  successfully handled the entering PiP transition
- Revoke TRUSTED_OVERLAY in WM once the task is not pinned any more

Flag: EXEMPT securityfix
Bug: 436270922
Test: atest PinnedStackTests
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:492b6ca4f31e79d9fced383a59957234aef8f8dc
Merged-In: Id849d93bc92209fc296ee91ef70b0ffe2a09550e
Change-Id: Id849d93bc92209fc296ee91ef70b0ffe2a09550e
---
 .../src/com/android/wm/shell/pip/PipTransition.java  | 12 ++++++++++--
 .../com/android/server/wm/RootWindowContainer.java   | 11 ++++++-----
 2 files changed, 16 insertions(+), 7 deletions(-)

diff --git a/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java b/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
index 9a73361a4..80693e201 100644
--- a/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
+++ b/libs/WindowManager/Shell/src/com/android/wm/shell/pip/PipTransition.java
@@ -521,8 +521,18 @@ public class PipTransition extends PipTransitionController {
             @NonNull Point leashOffset, @PipAnimationController.TransitionDirection int direction,
             @NonNull SurfaceControl.Transaction tx) {
         final boolean enteringPip = isInPipDirection(direction);
+        final SurfaceControl leash = mPipOrganizer.getSurfaceControl();
+        final boolean hasValidLeash = leash != null && leash.isValid();
         if (enteringPip) {
             mPipTransitionState.setTransitionState(ENTERED_PIP);
+            // TRUSTED_OVERLAY is granted iff Shell successfully receives the transition.
+            // It's revoked once the task exits pinned mode in
+            // RootWindowContainer#notifyActivityPipModeChanged
+            if (hasValidLeash) {
+                ProtoLog.d(ShellProtoLogGroup.WM_SHELL_PICTURE_IN_PICTURE,
+                        "Set TRUSTED_OVERLAY for Task#%d", taskInfo.taskId);
+                tx.setTrustedOverlay(leash, true);
+            }
         }
         // If we have an exit transition, but aren't playing a transition locally, it
         // means we're expecting the exit transition will be "merged" into another transition
@@ -530,8 +540,6 @@ public class PipTransition extends PipTransitionController {
         // the exit transition is merged.
         if ((mExitTransition == null || mMoveToBackTransition == null || isAnimatingLocally())
                 && mFinishCallback != null) {
-            final SurfaceControl leash = mPipOrganizer.getSurfaceControl();
-            final boolean hasValidLeash = leash != null && leash.isValid();
             WindowContainerTransaction wct = null;
             if (isOutPipDirection(direction)) {
                 // Only need to reset surface properties. The server-side operations were already
diff --git a/services/core/java/com/android/server/wm/RootWindowContainer.java b/services/core/java/com/android/server/wm/RootWindowContainer.java
index 1648141df..47bf2a637 100644
--- a/services/core/java/com/android/server/wm/RootWindowContainer.java
+++ b/services/core/java/com/android/server/wm/RootWindowContainer.java
@@ -2405,13 +2405,14 @@ class RootWindowContainer extends WindowContainer<DisplayContent>
             mService.getTaskChangeNotificationController().notifyActivityPinned(r);
         } else {
             mService.getTaskChangeNotificationController().notifyActivityUnpinned();
+            // Revoke the TRUSTED_OVERLAY here as a blanket policy.
+            if (task.getSurfaceControl() != null) {
+                mWmService.mTransactionFactory.get()
+                        .setTrustedOverlay(task.getSurfaceControl(), false)
+                        .apply();
+            }
         }
         mWindowManager.mPolicy.setPipVisibilityLw(inPip);
-        if (task.getSurfaceControl() != null) {
-            mWmService.mTransactionFactory.get()
-                    .setTrustedOverlay(task.getSurfaceControl(), inPip)
-                    .apply();
-        }
     }
 
     void executeAppTransitionForAllDisplay() {
-- 
2.52.0

